src/transformers/parallelization_utils.py

```diff
@@ -0,0 +1,1098 @@
+
+        if layer.bias is not None:
+            if slice_bias is True:
+                if layer.bias.dim() >= 1:
+                    bias = layer.bias.chunk(n_fused * world_size, dim=0)
+                    if n_fused > 1:
+                        bias = self._realign_fused_tensors(bias, world_size)
+                    layer.bias.data = bias[gpu_index].to(self.device)
+                else:
+                    layer.bias.data = layer.bias.to(self.device)
+            else:
+                layer.bias.data = layer.bias.to(self.device)
+
+        return layer
+
+    @staticmethod
+    def _realign_fused_tensors(tensor, world_size):
+        ranks = (len(tensor) + world_size - 1) // world_size
+        tensor = [tensor[i * world_size : (i + 1) * world_size] for i in range(ranks)]
+        tensor = list(map(lambda x: torch.cat([*x], dim=-1), zip(*tensor)))
+        return tensor
+
+    def _column_slice(self, layers, world_size, gpu_index):
+        return [
+            self._slice_layer(
+                layer=layer,
+                dim=0,
+                world_size=world_size,
+                gpu_index=gpu_index,
+                slice_bias=True,
+            )
+            for layer in layers
+        ]
+
+    def _row_slice(self, layers, world_size, gpu_index):
+        return [
+            self._slice_layer(
+                layer=layer,
+                dim=1,
+                world_size=world_size,
+                gpu_index=gpu_index,
+                slice_bias=False,
+            )
+            for layer in layers
+        ]
+
+    def _process_additional_layers(
+        self,
+        model,
+        mpu,
+        additional_layers,
+        config,
+    ):
+        additional_layers += self.policy.modelwise_copy_to_all(model, config)
+
+        for layer in additional_layers:
+            if layer.weight is not None:
+                layer.weight.data = layer.weight.to(self.device)
+            if layer.bias is not None:
+                layer.bias.data = layer.bias.to(self.device)
+            self._postprocess(layer, {"mpu": mpu})
+
+    def _postprocess(self, layer, attributes):
+        if layer.replace is not None:
+            for k, v in layer.replace.items():
+                for k_, v_ in attributes.items():
+                    setattr(k, k_, v_)
+                k.__class__ = v
+
+        if self.hold_params:
+            if layer.name in self.param_dict:
+                self.param_dict[layer.name].append(layer)
+            else:
+                self.param_dict[layer.name] = [layer]
+
+
+class ParallelizationMixin(object):
+    def _get_default_model_object(self):
+        raise NotImplementedError
+
+    def _get_parallelism_policy(self):
+        raise NotImplementedError
+
+    def _get_head_layers(self):
+        return []
+
+    def parallelize(
+        self,
+        tensor_model_parallel_size: int = 1,
+        pipeline_model_parallel_size: int = 1,
+        vocab_parallel_embedding: bool = None,
```

Comment:
Users can turn on and off vocab parallel embedding. but turn it on or off automatically by configuration.